tao_compiler/mlir/disc/transforms/disc_transpose_simplifier.cc

```diff
@@ -0,0 +1,851 @@
+  Value rhs = op->getOperand(1);
+  auto lhsDefiningOp = lhs.getDefiningOp();
+  auto rhsDefiningOp = rhs.getDefiningOp();
+
+  if (isConstOneBcast(lhsDefiningOp)) {
+    rewriter.replaceOp(op, {rhs});
+    return success();
+  }
+  if (isConstOneBcast(rhsDefiningOp)) {
+    rewriter.replaceOp(op, {lhs});
+    return success();
+  }
+  return failure();
+}
+
+void populateTransposeSimplifierPatterns(OwningRewritePatternList& patterns) {
+  patterns.insert(eliminateIdentityTranspse);
+  patterns.insert(propagateDimOfTranspse);
+  patterns.insert(propagateShapeOfTranspse);
+  patterns.insert(propagateIdentityMulOp);
+  // Adds canonicalization patterns to the list of patterns.
+  addCanonicalizationPatterns(patterns.getContext(), &patterns);
+}
+
+SmallVector<int64_t> getReversePermutation(ArrayRef<int64_t> perm) {
+  SmallVector<int64_t> newPerm(perm.size());
+  for (size_t i = 0; i < perm.size(); ++i) {
+    newPerm[perm[i]] = i;
+  }
+  return newPerm;
+}
+
+// Returns true is x and y are transpsoe ops and `y(x(INPUT)) = INPUT`.
+bool isMirroredTranspose(Operation* x, Operation* y) {
+  auto transposeX = dyn_cast_or_null<mhlo::TransposeOp>(x);
+  auto transposeY = dyn_cast_or_null<mhlo::TransposeOp>(y);
+  if (!transposeX || !transposeY) return false;
+
+  auto permXAttr = transposeX.permutation().getValues<int64_t>();
+  SmallVector<int64_t> permX{permXAttr.begin(), permXAttr.end()};
+  auto permYAttr = transposeY.permutation().getValues<int64_t>();
+  SmallVector<int64_t> permY{permYAttr.begin(), permYAttr.end()};
+
+  if (permX.size() != permY.size()) return false;
+
+  auto reverseX = getReversePermutation(permX);
+  return reverseX == permY;
+}
+
+RankedTensorType getTransposeOutputType(
+    Value value, const SmallVectorImpl<int64_t>& permutation, OpBuilder& b) {
+  // Compute the resulting shape.
+  llvm::SmallVector<int64_t, 4> transposed_shape;
+  ShapedType input_type = value.getType().cast<ShapedType>();
+  auto input_shape = input_type.getShape();
+  for (int64_t val : permutation) {
+    transposed_shape.push_back(input_shape[val]);
+  }
+  return RankedTensorType::get(transposed_shape, input_type.getElementType());
+}
+
+Operation* insertTranspose(Operation* op, Value value,
+                           const SmallVectorImpl<int64_t>& permutation,
+                           OpBuilder& b) {
+  auto permutationAttr = GetI64ElementsAttr(permutation, &b);
+
+  // Compute the resulting shape.
+  llvm::SmallVector<int64_t, 4> transposed_shape;
+  ShapedType input_type = value.getType().cast<ShapedType>();
+  auto input_shape = input_type.getShape();
+  for (auto val : permutation) {
+    transposed_shape.push_back(input_shape[val]);
+  }
+  auto transpose_type = getTransposeOutputType(value, permutation, b);
+  auto transpose_op = b.create<mhlo::TransposeOp>(op->getLoc(), transpose_type,
+                                                  value, permutationAttr);
+
+  if (auto attr = op->getAttr(placement_utils::kDiscPlaceAssignment))
```

Comment:
is it necessary for us to run Placer pass ahead of ConvRewriter & TransposeSimplifier?

Reply:
Currently, we use different layout for convOp on different devices.